File: cpp/TCALC.cpp
#include <math.h>
#include <ios>
#include "TCALC.h"

// OPerations
#define OP_PLUS          0
#define OP_MINUS         1
#define OP_MULTIPLY      2
#define OP_DIVIDE        3
#define OP_PERCENT       4
#define OP_POWER         5
#define OP_UMINUS        6
#define OP_SIN           7
#define OP_COS           8
#define OP_TG            9
#define OP_CTG           10
#define OP_ARCSIN        11
#define OP_ARCCOS        12
#define OP_ARCTG         13
#define OP_ARCCTG        14
#define OP_SH            15
#define OP_CH            16
#define OP_TH            17
#define OP_CTH           18
#define OP_EXP           19
#define OP_LG            20
#define OP_LN            21
#define OP_SQRT          22
#define OP_IN            23
#define OP_X             24

// Определени математических констант

#define M_PI             3.1415926535897932384626433832795
#define F_G              9.81
#define M_E              2.71828182845904523536
#define H_LEET           1337

TCALCNode *TCALC::CreateNode(double _value, TCALCNode *_left, TCALCNode *_right)
{
	TCALCNode *pNode = new TCALCNode(_value, _left, _right);
	return pNode;
}

void TCALC::SendError(int errNum) //Обработчик ошибок
{
	static char *errs[10] = 
	{ NULL,
	  NULL,
		"Внезапный конец выражения",
		"Конец выражения ожидается",
		"Пропущеннаи открывающая скобка",
		"Пропущенна закрывающая скобка",
		NULL
	};
	static char buffer[80];

	int len = strlen(curToken);



	if(*curToken=='\0')
		strcpy(curToken, "Пустое выражение");

	switch(errNum)
		{
			case 0:
				sprintf(buffer, "Что за хрень: '%s'", curToken);
				errs[0] = buffer;
				break;

			case 1:
				sprintf(buffer, "Что за хрень: '%s'", curToken);
				errs[1] = buffer;
				break;

			case 6:
				sprintf(buffer, "Что за хрень: '%s'", curToken);
				errs[6] = buffer;
				break;
		}

	TError error(errs[errNum], pos-len);

	root = NULL;

	throw error;

	return;
}

bool TCALC::GetToken(void)
{
	*curToken = '\0';

	while(expr[pos]==' ') pos++;

		if(expr[pos]=='\0')
		{
			curToken[0] = '\0';
			typToken = CALC_END;
			return true;
		}
	else
			if(IsDelim())
				{
					curToken[0] = expr[pos++];
					curToken[1] = '\0';
						switch(*curToken)
							{
								case '+':
									typToken = CALC_PLUS;
									return true;

								case '-':
									typToken = CALC_MINUS;
									return true;

								case '*':
									typToken = CALC_MULTIPLY;
									return true;

								case '/':
									typToken = CALC_DIVIDE;
									return true;

								case '%':
									typToken = CALC_PERCENT;
									return true;
								case '[': 
								case '(':
									typToken = CALC_L_BRACKET;
									return true;

								case ']': 

								case ')':
									typToken = CALC_R_BRACKET;
									return true;
							}
				}
	else if(IsLetter())
	{
		int i=0;
		while(IsLetter()) curToken[i++] = expr[pos++];
		curToken[i] = '\0';

		int len = strlen(curToken);
		for(i=0; i<len; i++)
			if(curToken[i]>='A' && curToken[i]<='Z')
				curToken[i] += 'a' - 'A';

		if(!strcmp(curToken, "x"))

			{
				typToken = CALC_X;
				return true;
			}

		else if(!strcmp(curToken, "leet"))

			{
				typToken = CALC_LEET;
				return true;
			}

		else if(!strcmp(curToken, "g")) 

			{
				typToken = CALC_G;
				return true;
			}




		else if(!strcmp(curToken, "pi"))
			
			{
				typToken = CALC_PI;
				return true;
			}

		else if(!strcmp(curToken, "sin"))
			{
				typToken = CALC_SIN;
				return true;
			}

		else if(!strcmp(curToken, "cos"))
		
			{
				typToken = CALC_COS;
				return true;
			}

		else if(!strcmp(curToken, "tg"))
			
			{
				typToken = CALC_TG;
				return true;
			}

		else if(!strcmp(curToken, "ctg"))
			
			{
				typToken = CALC_CTG;
				return true;
			}

		else if(!strcmp(curToken, "arcsin"))
			
			{
				typToken = CALC_ARCSIN;
				return true;
			}

		else if(!strcmp(curToken, "arccos"))
			
			{
				typToken = CALC_ARCCOS;
				return true;
			}


		else if(!strcmp(curToken, "arcctg"))
				
			{
				typToken = CALC_ARCCTG;
				return true;
			}



		else if(!strcmp(curToken, "sh"))
			
			{
				typToken = CALC_SH;
				return true;
			}

		else if(!strcmp(curToken, "ch"))

			{
				typToken = CALC_CH;
				return true;
			}

		else if(!strcmp(curToken, "th"))
			
			{	
				typToken = CALC_TH;
				return true;
			}

		else if(!strcmp(curToken, "cth"))
			
			{
				typToken = CALC_CTH;
				return true;
			}

		else if(!strcmp(curToken, "exp"))
			
			{
				typToken = CALC_EXP;
				return true;
			}

		else if(!strcmp(curToken, "lg"))
			
			{
				typToken = CALC_LG;
				return true;
			}

		else if(!strcmp(curToken, "ln"))
			
			{
				typToken = CALC_LN;
				return true;
			}

		else if(!strcmp(curToken, "sqrt"))
		
			{
				typToken = CALC_SQRT;
				return true;
			}

		else SendError(0);
	}






	else if(IsDigit() || IsPoint())
	{
		int i=0;
		while(IsDigit()) curToken[i++] = expr[pos++];
		if(IsPoint())
			{
				curToken[i++] = expr[pos++];
				while(IsDigit()) curToken[i++] = expr[pos++];
			}
		curToken[i] = '\0';
		typToken = CALC_NUMBER;
		return true;
	}

	else
	{
		curToken[0] = expr[pos++];
		curToken[1] = '\0';
		SendError(1);
	}

	return false;
}      

bool TCALC::Compile(char *_expr)
{

	pos = 0;
	expr = _expr;
	*curToken = '\0';
	if(root!=NULL)
		{
			DelTree(root);
			root = NULL;
		}

	GetToken();
	if(typToken==CALC_END) SendError(2);

	root = Expr();
	if(typToken!=CALC_END) SendError(3);
	return true;
}

TCALCNode *TCALC::Expr(void)
{
	TCALCNode *temp = Expr1();

	while(1)
	{
		if(typToken == CALC_PLUS)
			{
				GetToken();
				temp = CreateNode(OP_PLUS, temp, Expr1());
			}

		else if(typToken==CALC_MINUS)
			{
				GetToken();
				temp = CreateNode(OP_MINUS, temp, Expr1());
			}
		else break;
	}


	return temp;
}



TCALCNode *TCALC::Expr1(void)
{
	TCALCNode *temp = Expr2();

	while(1)
	{
		if(typToken==CALC_MULTIPLY)
			{
				GetToken();
				temp = CreateNode(OP_MULTIPLY, temp, Expr2());
			}

		else if(typToken==CALC_DIVIDE)
			{
				GetToken();
				temp = CreateNode(OP_DIVIDE, temp, Expr2());
			}

		else if(typToken==CALC_PERCENT)
			{
				GetToken();
				temp = CreateNode(OP_PERCENT, temp, Expr2());
			}
		else break;
	}

	return temp;
}

TCALCNode *TCALC::Expr2(void)
{
    TCALCNode *temp = Expr3();
    return temp;
}

TCALCNode *TCALC::Expr3(void)
{
	TCALCNode *temp;

	if(typToken==CALC_PLUS)
	{
		GetToken();
		temp = Expr4();
	}


	else if(typToken==CALC_MINUS)
	{
		GetToken();
		temp = CreateNode(OP_UMINUS, Expr4());
	}

	else
		temp = Expr4();

	return temp;      
}

TCALCNode *TCALC::Expr4(void)
{
	TCALCNode *temp;

	if(typToken >= CALC_SIN && typToken<=CALC_X)
		{
			temp = CreateNode(OP_SIN-CALC_SIN+typToken);
			GetToken();

			if(typToken!=CALC_L_BRACKET)
				SendError(4);
				GetToken();

			temp->left = Expr();

			if(typToken!=CALC_R_BRACKET)
				SendError(5);
				GetToken();
		}

	else
		temp = Expr5();

	return temp;
}

TCALCNode *TCALC::Expr5(void)
{
	TCALCNode *temp;

	switch(typToken)
		{
			case CALC_NUMBER:
				temp = CreateNode((double)atof(curToken));
				GetToken();
                break;
			case CALC_PI:
				temp = CreateNode((double)M_PI);
				GetToken();
				break;
			case CALC_G:
				temp =CreateNode((double)F_G);
				GetToken();
				break;
			case CALC_L_BRACKET:
				GetToken();
				temp = Expr();
				if(typToken!=CALC_R_BRACKET) SendError(5);
				GetToken();
				break;
			case CALC_LEET:
				temp = CreateNode((double) H_LEET);
				GetToken();
				break;
			default:
				SendError(6);
	}

	return temp;         
}

double TCALC::Evaluate(void)
{
	result = CalcTree(root);
	return result;
}

double TCALC::CalcTree(TCALCNode *tree) // Обработчик действий
{

	static double temp;

	if(tree->left==NULL && tree->right==NULL)
		return tree->value;  
	
	else
		switch((int)tree->value)
	{
		case OP_PLUS:
			return CalcTree(tree->left)+CalcTree(tree->right);
			// "(tree->left)+(tree->right)"

		case OP_MINUS:
			return CalcTree(tree->left)-CalcTree(tree->right);
			// "(tree->left)-(tree->right)"

		case OP_MULTIPLY:
			return CalcTree(tree->left)*CalcTree(tree->right);
			// "(tree->left)*(tree->right)"

		case OP_DIVIDE:
			return CalcTree(tree->left)/CalcTree(tree->right);	
			// "(tree->left)/(tree->right)"

		case OP_PERCENT:
			return (int)CalcTree(tree->left)%(int)CalcTree(tree->right);	
			// "(tree->left)%(tree->right)"

		case OP_POWER:
			return (double)pow(CalcTree(tree->left),CalcTree(tree->right));
			// "(tree->left)^(tree->right)"

		case OP_UMINUS:
			return -CalcTree(tree->left);
			// "-(tree->left)"

		case OP_SIN:
			return sin(CalcTree(tree->left));
			// "sin(tree->left"

		case OP_COS:
			return cos(CalcTree(tree->left));
			// "cos(tree->left)"

		case OP_TG:
			return tan(CalcTree(tree->left));
			// "tan(tree->left)"

		case OP_CTG:
			return 1.0/tan(CalcTree(tree->left));
			// "ctg(tree->left)"

		case OP_ARCSIN:
			return asin(CalcTree(tree->left));
			// "arcsin(tree->left)"

		case OP_ARCCOS:
			return acos(CalcTree(tree->left));
			// "arccos(tree->left)"

		case OP_ARCTG:
			return atan(CalcTree(tree->left));
			// "atctg(tree->left)"

		case OP_ARCCTG:
			return M_PI/2.0-atan(CalcTree(tree->left));
			// "arcctg(tree->left)"

		case OP_SH:
			temp = CalcTree(tree->left);
			return (exp(temp)-exp(-temp))/2.0;	
			// "sh(tree->left)"

		case OP_CH:
			temp = CalcTree(tree->left);
			return (exp(temp)+exp(-temp))/2.0;
			// "ch(tree->left)"

		case OP_TH:
			temp = CalcTree(tree->left);
			return (exp(temp)-exp(-temp))/(exp(temp)+exp(-temp));
			// "th(tree->left)"

		case OP_CTH:
			temp = CalcTree(tree->left);
			return (exp(temp)+exp(-temp))/(exp(temp)-exp(-temp));
			// "cth(tree->left)"

		case OP_EXP:
			return exp(CalcTree(tree->left));
			// "e^(tree->left)"

		case OP_LG:
			return log10(CalcTree(tree->left));
			// "log10(tree->left)"

		case OP_LN:
			return log(CalcTree(tree->left));
			// "loge(tree->left)"

		case OP_SQRT:
			return sqrt(CalcTree(tree->left));
			// "(tree->left)^(1/2)"

		case OP_X:	
			return x[int(CalcTree(tree->left))];

		case OP_IN:
			return 1;

	}

	return 0;
}

void TCALC::DelTree(TCALCNode *tree)
{
	if(tree==NULL) return;

	DelTree(tree->left);
	DelTree(tree->right);

	delete tree;

	return;
}

File: python/tcalc/tcalc.py
import math
from typing import List, Union

# Math consts
M_PI = 3.1415926535897932384626433832795
F_G = 9.81
M_E = 2.71828182845904523536
H_LEET = 1337

# Defines
MAX_EXPR_LEN = 255
MAX_TOKEN_LEN = 80

# Brackets
CALC_END = -1
CALC_L_BRACKET = -2
CALC_R_BRACKET = -3
CALC_NUMBER = -4

# Operations
OP_PLUS = 0
OP_MINUS = 1
OP_MULTIPLY = 2
OP_DIVIDE = 3
OP_PERCENT = 4
OP_POWER = 5
OP_UMINUS = 6

# Math Operations
OP_SIN = 7
OP_COS = 8
OP_TG = 9
OP_CTG = 10
OP_ARCSIN = 11
OP_ARCCOS = 12
OP_ARCTG = 13
OP_ARCCTG = 14
OP_SH = 15
OP_CH = 16
OP_TH = 17
OP_CTH = 18
OP_EXP = 19
OP_LG = 20
OP_LN = 21
OP_SQRT = 22
OP_IN = 23
CALC_PI = 24
CALC_G = 25
CALC_LEET = 26

TERMINATOR = "\x00"


def str_len(list1):
    list1 = list(list1)
    i = 0
    while list1[i] != TERMINATOR:
        i = i + 1
    return i


def strcmp(list1, list2):
    list1 = list(list1)
    list2 = list(list2)
    total = len(list1) if len(list1) < len(list2) else len(list2)
    for i in range(0, total):
        if i > len(list1) - 1:
            return not True
        elif i > len(list2):
            return not True
        elif list1[i] != list2[i]:
            return not False
        else:
            continue


def atof(text: List[str]):
    text = "".join(text[:str_len(text)])
    integer = int(text)
    double = float(text)
    if integer == double:
        return integer
    else:
        return double


class TCalcException(Exception):

    def __init__(self, reason: str):
        self.reason = reason


    def __str__(self):
        return self.reason


class TCalcNode:

    def __init__(self, _value=0, _left=None, _right=None):
        self.left = _left
        self.right = _right
        self.value = _value


class TCalc:
    expr_ = [TERMINATOR for i in range(0, MAX_EXPR_LEN)]
    cur_token = [TERMINATOR for i in range(0, MAX_TOKEN_LEN)]

    def __init__(self):
        self.root = None
        self.typ_token = 0
        self.pos = 0
        self.result = 0

    @classmethod
    def create_node(cls, _value, _left, _right) -> TCalcNode:
        return TCalcNode(_value, _left, _right)

    def is_delim(self):
        for char in list("+-*/%^()[]"):
            if self.expr_[self.pos] == char:
                return True
            else:
                continue
        return False

    def is_letter(self):
        return (
            (ord('a') <= ord(self.expr_[self.pos]) <= ord('z')) or (ord('A') <= ord(self.expr_[self.pos]) <= ord('Z'))
        )

    def is_digit(self):
        return ord('0') <= ord(self.expr_[self.pos]) <= ord('9')

    def is_point(self):
        return self.expr_[self.pos] == "."

    def get_token(self):
        self.cur_token[0] = TERMINATOR
        while self.expr_[self.pos] == " ":
            self.pos = self.pos + 1
        if self.expr_[self.pos] == TERMINATOR:
            self.cur_token[0] = TERMINATOR
            self.typ_token = CALC_END
            return True
        elif self.is_delim():
            self.cur_token[0] = self.expr_[self.pos]
            self.pos = self.pos + 1
            self.cur_token[1] = TERMINATOR
            tmp = "".join(self.cur_token[:str_len(self.cur_token)])
            if tmp == "+":
                self.typ_token = OP_PLUS
                return True
            elif tmp == "-":
                self.typ_token = OP_MINUS
                return True
            elif tmp == "^":
                self.typ_token = OP_POWER
                return True
            elif tmp == "*":
                self.typ_token = OP_MULTIPLY
                return True
            elif tmp == "/":
                self.typ_token = OP_DIVIDE
                return True
            elif tmp == "%":
                self.typ_token = OP_PERCENT
                return True
            elif tmp == "[" or tmp == "(":
                self.typ_token = CALC_L_BRACKET
                return True
            elif tmp == "]" or tmp == ")":
                self.typ_token = CALC_R_BRACKET
                return True
        elif self.is_letter():
            i = 0
            while self.is_letter():
                self.cur_token[i] = self.expr_[self.pos]
                self.pos = self.pos + 1
                i = i + 1
            self.cur_token[i] = TERMINATOR
            _len = str_len(self.cur_token)
            for i in range(0, _len):
                if ord('A') <= ord(self.cur_token[i]) <= ord('Z'):
                    self.cur_token[i] = chr(ord(self.cur_token[i]) + ord('a') - ord('A'))
            if not strcmp(self.cur_token, list("leet")):
                self.typ_token = CALC_LEET
                return True
            elif not strcmp(self.cur_token, list("g")):
                self.typ_token = CALC_G
                return True
            elif not strcmp(self.cur_token, list("pi")):
                self.typ_token = CALC_PI
                return True
            elif not strcmp(self.cur_token, list("sin")):
                self.typ_token = OP_SIN
                return True
            elif not strcmp(self.cur_token, list("cos")):
                self.typ_token = OP_COS
                return True
            elif not strcmp(self.cur_token, list("tg")):
                self.typ_token = OP_TG
                return True
            elif not strcmp(self.cur_token, list("ctg")):
                self.typ_token = OP_CTG
                return True

            elif not strcmp(self.cur_token, list("arcsin")):
                self.typ_token = OP_ARCSIN
                return True
            elif not strcmp(self.cur_token, list("arccos")):
                self.typ_token = OP_ARCCOS
                return True
            elif not strcmp(self.cur_token, list("arctg")):
                self.typ_token = OP_ARCTG
                return True
            elif not strcmp(self.cur_token, list("arcctg")):
                self.typ_token = OP_ARCCTG
                return True
            elif not strcmp(self.cur_token, list("sh")):
                self.typ_token = OP_SH
                return True
            elif not strcmp(self.cur_token, list("ch")):
                self.typ_token = OP_CH
                return True
            elif not strcmp(self.cur_token, list("th")):
                self.typ_token = OP_TH
                return True
            elif not strcmp(self.cur_token, list("cth")):
                self.typ_token = OP_CTH
                return True
            elif not strcmp(self.cur_token, list("exp")):
                self.typ_token = OP_EXP
                return True
            elif not strcmp(self.cur_token, list("lg")):
                self.typ_token = OP_LG
                return True
            elif not strcmp(self.cur_token, list("ln")):
                self.typ_token = OP_LN
                return True
            elif not strcmp(self.cur_token, list("sqrt")):
                self.typ_token = OP_SQRT
                return True
            else:
                self.send_error(0)
        elif self.is_digit() or self.is_point():
            i = 0
            while self.is_digit():
                self.cur_token[i] = self.expr_[self.pos]
                self.pos = self.pos + 1
                i = i + 1
            if self.is_point():
                self.cur_token[i] = self.expr_[self.pos]
                self.pos = self.pos + 1
                i = i + 1
                while self.is_digit():
                    self.cur_token[i] = self.expr_[self.pos]
                    self.pos = self.pos + 1
                    i = i + 1
            self.cur_token[i] = TERMINATOR
            self.typ_token = CALC_NUMBER
            return True
        else:
            self.cur_token[0] = self.expr_[self.pos]
            self.pos = self.pos + 1
            self.cur_token[1] = TERMINATOR
            self.send_error(1)
        return False

    def expr(self):
        temp = self.expr1()
        while True:
            if self.typ_token == OP_PLUS:
                self.get_token()
                temp = self.create_node(OP_PLUS, temp, self.expr1())
            elif self.typ_token == OP_MINUS:
                self.get_token()
                temp = self.create_node(OP_MINUS, temp, self.expr1())
            else:
                break
        return temp

    def expr1(self):
        temp = self.expr2()
        while True:
            if self.typ_token == OP_MULTIPLY:
                self.get_token()
                temp = self.create_node(OP_MULTIPLY, temp, self.expr2())
            elif self.typ_token == OP_DIVIDE:
                self.get_token()
                temp = self.create_node(OP_DIVIDE, temp, self.expr2())
            elif self.typ_token == OP_PERCENT:
                self.get_token()
                temp = self.create_node(OP_PERCENT, temp, self.expr2())
            else:
                break
        return temp

    def expr2(self):
        if self.typ_token == OP_PLUS:
            self.get_token()
            temp = self.expr3()
        elif self.typ_token == OP_MINUS:
            self.get_token()
            temp = self.create_node(OP_UMINUS, self.expr3(), None)
        else:
            temp = self.expr3()
        return temp

    def expr3(self):
        if OP_SIN <= self.typ_token <= OP_SQRT + 1:
            temp = self.create_node(OP_SIN - OP_SIN + self.typ_token, None, None)
            self.get_token()
            if self.typ_token != CALC_L_BRACKET:
                self.send_error(4)
            self.get_token()
            temp.left = self.expr()
            if self.typ_token != CALC_R_BRACKET:
                self.send_error(5)
            self.get_token()
        else:
            temp = self.expr4()
        return temp

    def expr4(self):
        temp = None
        if self.typ_token == CALC_NUMBER:
            temp = self.create_node(atof(self.cur_token), None, None)
            self.get_token()
        elif self.typ_token == CALC_PI:
            temp = self.create_node(M_PI, None, None)
            self.get_token()
        elif self.typ_token == CALC_G:
            temp = self.create_node(F_G, None, None)
            self.get_token()
        elif self.typ_token == CALC_L_BRACKET:
            self.get_token()
            temp = self.expr()
            if self.typ_token != CALC_R_BRACKET:
                self.send_error(5)
            self.get_token()
        elif self.typ_token == CALC_LEET:
            temp = self.create_node(H_LEET, None, None)
            self.get_token()
        else:
            self.send_error(5)
        return temp

    def send_error(self, err_num: int):
        if self.cur_token == TERMINATOR:
            raise TCalcException("Пустое выражение")
        elif err_num == 2:
            raise TCalcException("Внезапный конец выражения")
        elif err_num == 3:
            raise TCalcException("Конец выражения ожидается")
        elif err_num == 4:
            raise TCalcException("Пропущена открывающая скобка")
        elif err_num == 5:
            raise TCalcException("Пропущена закрывающая скобка")
        else:
            raise TCalcException("Неизвестная ошибка")

    def calc_tree(self, tree):
        if tree.left is None and tree.right is None:
            return tree.value
        else:
            op = tree.value
            if op == OP_PLUS:
                return self.calc_tree(tree.left) + self.calc_tree(tree.right)
            elif op == OP_MINUS:
                return self.calc_tree(tree.left) - self.calc_tree(tree.right)
            elif op == OP_MULTIPLY:
                return self.calc_tree(tree.left) * self.calc_tree(tree.right)
            elif op == OP_DIVIDE:
                return self.calc_tree(tree.left) / self.calc_tree(tree.right)
            elif op == OP_PERCENT:
                return self.calc_tree(tree.left) % (self.calc_tree(tree.right))
            elif op == OP_POWER:
                return math.pow(self.calc_tree(tree.left), self.calc_tree(tree.right))
            elif op == OP_UMINUS:
                return -self.calc_tree(tree.left)
            elif op == OP_SIN:
                return math.sin(self.calc_tree(tree.left))
            elif op == OP_COS:
                return math.cos(self.calc_tree(tree.left))
            elif op == OP_TG:
                return math.tan(self.calc_tree(tree.left))
            elif op == OP_CTG:
                return 1.0 / math.tan(self.calc_tree(tree.left))
            elif op == OP_ARCSIN:
                return math.asin(self.calc_tree(tree.left))
            elif op == OP_ARCCOS:
                return math.acos(self.calc_tree(tree.left))
            elif op == OP_ARCTG:
                return math.atan(self.calc_tree(tree.left))
            elif op == OP_ARCCTG:
                return M_PI / 2.0 - math.atan(self.calc_tree(tree.left))
            elif op == OP_SH:
                temp = self.calc_tree(tree.left)
                return (math.exp(temp) - math.exp(-temp)) / 2.0
            elif op == OP_CH:
                temp = self.calc_tree(tree.left)
                return (math.exp(temp) + math.exp(-temp)) / 2.0
            elif op == OP_TH:
                temp = self.calc_tree(tree.left)
                return (math.exp(temp) - math.exp(-temp)) / (math.exp(temp) + math.exp(-temp))
            elif op == OP_CTH:
                temp = self.calc_tree(tree.left)
                return (math.exp(temp) + math.exp(-temp)) / (math.exp(temp) - math.exp(-temp))
            elif op == OP_EXP:
                return math.exp(self.calc_tree(tree.left))
            elif op == OP_LG:
                return math.log10(self.calc_tree(tree.left))
            elif op == OP_LN:
                return math.log(self.calc_tree(tree.left))
            elif op == OP_SQRT:
                return math.sqrt(self.calc_tree(tree.left))
            elif op == OP_IN:
                return 1
        return 0

    def evalute(self, _expr: Union[List[str], str]):
        _expr = list(_expr) if isinstance(_expr, str) else _expr
        self.root = None
        self.typ_token = 0
        self.pos = 0
        self.result = 0

        self.expr_ = _expr + [TERMINATOR]
        self.get_token()
        if self.typ_token == CALC_END:
            self.send_error(2)
        self.root = self.expr()
        if self.typ_token != CALC_END:
            self.send_error(3)

        return self.calc_tree(self.root)
